### PR TITLE
Provider, Resource, and DataSource types

### LINKED ISSUE #14: Define provider type.
### Module version

```
N/A
```

### Use-cases

We want provider developers to be able to define their provider. This needs to be able to store information on the provider configuration's schema, the resources and data sources registered to the provider, and optionally the provider_meta configuration schema. It also needs a way to handle the `ConfigureProvider` RPC that passes the provider configuration.

### Proposal

A `Provider` interface that provider developers can implement.

### References

* [structs/interfaces design docs](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/structs-interfaces.md)
* [factories design docs](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/factories.md)

### LINKED ISSUE #15: Define resource and data source types
### Module version

```
N/A
```

### Use-cases

We want provider developers to be able to define new resources and data sources in their provider. They'll need to able to represent their schema, and to implement CRUD methods that create, read, update, and destroy the resource as needed in response to the `ApplyResourceChange` RPC. Additionally, they may optionally want to modify the plan as part of `PlanResourceChange`. They'll need access to the configuration data from the `ConfigureProvider` RPC stored on the provider type defined in #14.

### Proposal

`Resource` and `DataSource` interfaces that provider developers can implement.

### References

* [structs/interfaces design docs](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/structs-interfaces.md)
* [factories design docs](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/factories.md)

### PR DESCRIPTION
**Pull request by kmoe**
closes #14
closes #15

### CODE REVIEW

File: provider.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+)
+
+// Provider is the core interface that all Terraform providers must implement.
+type Provider interface {
+	// GetSchema returns the schema for this provider's configuration. If
+	// this provider has no configuration, return nil.
+	GetSchema(context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
```
Comment (paddycarver) [inline provider.go:13]:
```suggestion
	GetSchema(context.Context) (schema.Schema, []*tfprotov6.Diagnostic)
```

File: provider.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	// GetDataSources returns a map of the data source types this provider
+	// supports.
+	GetDataSources(context.Context) (map[string]DataSourceType, []*tfprotov6.Diagnostic)
+}
+
+// ProviderWithProviderMeta is a provider with a provider meta schema.
+// This functionality is currently experimental and subject to change or break
+// without warning; it should only be used by providers that are collaborating
+// on its use with the Terraform team.
+type ProviderWithProviderMeta interface {
```
Comment (paddycarver) [inline provider.go:37]: Should this inline the provider type, to be a superset of it?

```suggestion
type ProviderWithProviderMeta interface {
    Provider
```

File: resource.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// A ResourceType is a type of resource. For each type of resource this provider
+// supports, it should instantiate a struct implementing ResourceType and return
```
Comment (paddycarver) [inline resource.go:11]:
```suggestion
// supports, it should define a type implementing ResourceType and return
```

File: resource.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// it as a value in the map returned by Provider.GetResources.
+type ResourceType interface {
+	// GetSchema returns the schema for this resource.
+	GetSchema(context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
```
Comment (paddycarver) [inline resource.go:15]:
```suggestion
	GetSchema(context.Context) (schema.Schema, []*tfprotov6.Diagnostic)
```

**Review by paddycarver — CHANGES_REQUESTED** [review]
Some minor nitpicks and suggestions, but this looks pretty much right to me.

File: data_source.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package tfsdk
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+)
+
+// A DataSourceType is a type of data source. For each type of data source this
+// provider supports, it should instantiate a struct implementing DataSourceType
```
[the remark this answers is not in the document]
Comment (kmoe) [reply data_source.go:10]: If we clarify in this direction, I'd rather we change the whole thing to "it should define a type implementing DataSourceType and return an instance [of that type] in the map returned by Provider.GetDataSources." Will do so and we can see if it reads better.

**Pull request merged**